postgrest/_sync/request_builder.py

```diff
@@ -59,13 +59,85 @@ def execute(self) -> APIResponse:
         )
 
         try:
-            return APIResponse.from_http_request_response(r)
+            if 200 <= r.status_code <= 299:  # Response.ok from JS (https://developer.mozilla.org/en-US/docs/Web/API/Response/ok)
+                return APIResponse.from_http_request_response(r)
+            else:
+                raise APIError(r.json())
         except ValidationError as e:
             raise APIError(r.json()) from e
 
 
+class SyncMaybeSingleRequestBuilder(SyncQueryRequestBuilder):
+    def execute(self) -> APIResponse:
+        try:
+            r = super().execute()
+        except APIError as e:
+            if e.details and "Results contain 0 rows" in e.details:
+                return APIResponse.from_dict({
+                    "data": None,
+                    "error": None,
+                    "count": 0  # NOTE: needs to take value from res.count
+                })
+        return r
+
+
+class SyncQueryBuilder:
```

Comment:
Does this class only add the functionality of checking whether `single` or `maybe_single` was used?

Reply:
The naming of our query builder classes have gotten so ambiguous haha

Reply:
This class is more like a factory class I'd say. On the naming system, that's true, any suggestions?

Reply:
Following [the factory design pattern[(https://refactoring.guru/design-patterns/factory-method), changing the parent's name as `AsyncQueryFactory`.